# Capstone Project - Interim Report

# 1) Industry Review
### a) Industry Review – Current practices, Background Research 
### b) Literature Survey - Publications, Application, past and undergoing research

# 2) Dataset and Domain

### Data Dictionary

| Sl No | Variable Name	| Role | Type | Demographic | Description | Units | Missing Values | 
| ------  | ------  |   ------  |  ------  |  ------  |  ------  |  ------  |  ------  | 
| 1 | encounter_id | ID |  |   | Unique identifier of an encounter |  | no | 
| 2 | patient_nbr | ID |  |   | Unique identifier of a patient  |  | no | 
| 3 | race | Feature | Categorical | Race | Values: Caucasian, Asian, African American, Hispanic, and other  |  | yes |
| 4 | gender | Feature| Categorical| Gender| Values: male, female, and unknown/invalid  |  | no |
| 5 | age | Feature | Categorical | Age | Grouped in 10-year intervals: [0, 10), [10, 20),..., [90, 100)   |  | no | 
| 6 | weight| Feature| Categorical|  |  Weight in pounds.|  |  yes| 
| 7 | admission_type_id| Feature| Categorical |  |  Integer identifier corresponding to 9 distinct values, for example, emergency, urgent, elective, newborn, and not available |  | no | 
| 8 | discharge_disposition_id | Feature| Categorical |  | Integer identifier corresponding to 29 distinct values, for example, discharged to home, expired, and not available  | | no | 
| 9 | admission_source_id | Feature | Categorical |  | Integer identifier corresponding to 21 distinct values, for example, physician referral, emergency room, and transfer from a hospital  |  | no | 
| 10 | time_in_hospital | Feature | Integer |  | Integer   number of days between admission and discharge  | | no  | 
| 11 | payer_code | Feature | Categorical |  | Integer  identifier corresponding to 23 distinct values, for example, Blue Cross/Blue Shield, Medicare, and self-pay |  | yes  | 
| 12 | medical_specialty | Feature | Categorical |  | Integer   identifier of a specialty of the admitting physician, corresponding to 84 distinct values, for example, cardiology, internal medicine, family/general practice, and surgeon |  | yes |
| 13 | num_lab_procedures | Feature | Integer | |  Number of lab tests performed during the encounter |  | no |
| 14 | num_procedures | Feature | Integer |  | Number of procedures (other than lab tests) performed during the encounter |   |  no | 
| 15 | num_medications | Feature | Integer  |  | Number of distinct generic names administered during the encounter |   |  no |  
| 16 | number_outpatient | Feature | Integer  |  | Number of outpatient visits of the patient in the year preceding the encounter |  | no| 
| 17 | number_emergency | Feature | Integer |  |  Number of emergency visits of the patient in the year preceding the encounter |  | no| 
| 18 | number_inpatient | Feature | Integer |  | Number of inpatient visits of the patient in the year preceding the encounter |  | no| 
| 19 | diag_1 | Feature | Categorical |  |  The primary diagnosis (coded as first three digits of ICD9); 848 distinct values |  | yes | 
| 20 | diag_2 | Feature | Categorical |  |  Secondary diagnosis (coded as first three digits of ICD9); 923 distinct values|  | yes| 
| 21 | diag_3 | Feature | Categorical |  |  Additional secondary diagnosis (coded as first three digits of ICD9); 954 distinct values |  | yes | 
| 22 | number_diagnoses | Feature | Integer | |  Number of diagnoses entered to the system	 |  | no | 
| 23 | max_glu_serum | Feature | Categorical | |  Indicates the range of the result or if the test was not taken. Values: >200, >300, normal, and none if not measured |  | no | 
| 24 | A1Cresult | Feature | Categorical | |  Indicates the range of the result or if the test was not taken. Values: >8 if the result was greater than 8%, >7 if the result was greater than 7% but less than 8%, normal if the result was less than 7%, and none if not measured. |  | no | 
| 25 | metformin | Feature | Categorical | |  The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed	 |  | no | 
| 26 | repaglinide | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 27 | nateglinide | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 28 | chlorpropamide | Feature	 | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 29 | glimepiride | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 30 | acetohexamide | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 31 | glipizide | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 32 | glyburide | 	Feature | 	Categorical	 |  | 	The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 33 | tolbutamide | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 34 | pioglitazone | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 35 | rosiglitazone | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 36 | acarbose | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 37 | miglitol | Feature | Categorical	 |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 38 | troglitazone | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 39 | tolazamide | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 40 | examide | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 41| citoglipton | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 42 | insulin | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 43 | glyburide-metformin | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 44 | glipizide-metformin | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 45 | glimepiride-pioglitazone | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 46 | metformin-rosiglitazone | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 47 | metformin-pioglitazone | Feature | Categorical |  | The feature indicates whether the drug was prescribed or there was a change in the dosage. Values: up if the dosage was increased during the encounter, down if the dosage was decreased, steady if the dosage did not change, and no if the drug was not prescribed |  | no | 
| 48 | change | Feature | Categorical |  | Indicates if there was a change in diabetic medications (either dosage or generic name). Values: change and no change |  | no | 
| 49 | iabetesMed | Feature | Categorical |  | Indicates if there was any diabetic medication prescribed. Values: yes and no |  | no | 
| 50 | readmitted | Target | Categorical |  | Days to inpatient readmission. Values: <30 if the patient was readmitted in less than 30 days, >30 if the patient was readmitted in more than 30 days, and No for no record of readmission. |  | no | 

###	Alternate sources of data that can supplement the core dataset (at least 2-3 columns)
* 1) Incomplete discharge planning
* 2) Complex care needs of senior citizens

Reference: 
https://pmc.ncbi.nlm.nih.gov/articles/PMC4953449

### Project Justification -  Project Statement, Complexity involved, Project Outcome –Commercial, Academic or Social value

### Project Statement

### Complexity involved: *Medium*
### Project Outcome –Commercial, Academic or Social value: *Commercial and Social value*

### 1) Importing the required libraries

In [11]:
from datetime import datetime
from datetime import timedelta
import time

In [12]:
# check scikit-learn version
import sklearn
print(sklearn.__version__)

1.5.1


In [13]:
import warnings
warnings.filterwarnings("ignore")

In [14]:
import   pandas                 as      pd
import   numpy                  as      np
import   matplotlib.pyplot      as      plt
import   seaborn                as      sns
import   scipy.stats            as      stats
from     scipy                  import  stats
import   statsmodels.api        as     sm
from     sklearn.preprocessing                 import   LabelEncoder
from     statsmodels.stats.outliers_influence  import   variance_inflation_factor
import   warnings


from     sklearn.metrics          import   accuracy_score, explained_variance_score

from     sklearn.experimental     import   enable_iterative_imputer
from     sklearn.impute           import   IterativeImputer
from     sklearn.preprocessing    import   RobustScaler

from   sklearn.feature_selection   import  RFECV
from   sklearn.feature_selection   import  RFE
from   sklearn                     import  metrics
from   sklearn.pipeline            import  Pipeline
from   sklearn                     import  preprocessing   
from   sklearn.model_selection     import  cross_val_score
from   sklearn.model_selection     import  RepeatedStratifiedKFold

from   sklearn.model_selection     import  train_test_split

In [15]:
from   sklearn.experimental        import enable_iterative_imputer
from   sklearn.impute              import IterativeImputer

In [16]:
from   sklearn.linear_model        import   LogisticRegression, LinearRegression
from   sklearn.tree                import   DecisionTreeRegressor
from   sklearn.tree                import   DecisionTreeClassifier
from   sklearn.neighbors           import   KNeighborsClassifier
from   sklearn.naive_bayes         import   GaussianNB
from   sklearn.ensemble            import   RandomForestClassifier, AdaBoostClassifier
import xgboost                     as       xgb

In [17]:
def missing_zero_values_table(df):
        zero_val = (df == 0.00).astype(int).sum(axis=0)
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mz_table = pd.concat([mis_val, mis_val_percent], axis=1)
        mz_table = mz_table.rename(
        columns = {0 : 'Count of Missing Values', 1 : '% Missing Values'})
        mz_table['% Missing Values'] = 100 * mz_table['Count of Missing Values'] / len(df)
        mz_table['Data Type'] = df.dtypes
        mz_table = mz_table[
            mz_table.iloc[:,1] != 0].sort_values(
        '% Missing Values', ascending=False).round(2)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns and " + str(df.shape[0]) + " Rows.\n"      
            "There are " + str(mz_table.shape[0]) +
              " columns that have missing values.")
        return mz_table

In [18]:
def rev_dict(d1):
    return dict(map(lambda key: (d1[key], key), d1.keys()))

In [19]:
d1 =   {0: 'A1', 1: 'A10', 2: 'A11', 3: 'A12', 4: 'A2', 5: 'A3', 6: 'A4', 7: 'A5', 8: 'A6', 9: 'A7', 10: 'A8', 11: 'A9'}
print(dict(map(lambda key: (d1[key], key), d1.keys())))

{'A1': 0, 'A10': 1, 'A11': 2, 'A12': 3, 'A2': 4, 'A3': 5, 'A4': 6, 'A5': 7, 'A6': 8, 'A7': 9, 'A8': 10, 'A9': 11}


In [20]:
rev_dict(d1)

{'A1': 0,
 'A10': 1,
 'A11': 2,
 'A12': 3,
 'A2': 4,
 'A3': 5,
 'A4': 6,
 'A5': 7,
 'A6': 8,
 'A7': 9,
 'A8': 10,
 'A9': 11}

In [21]:
def chk_constant_rows(df):
    colsn       =   df.columns
    uniq_cols   =   list()
    for  col in colsn:
        counts = len(pd.unique(df[col]))
        if counts == 1:
          print("\nUnique Column: {0} Count {1}".format(col, counts))
          uniq_cols.append(col)
    return uniq_cols

In [22]:
def label(df, LABEL_COL):
    from    sklearn.preprocessing import   LabelEncoder
    import  pandas                as       pd
    le        =   LabelEncoder()
    _df       =   df.copy()
    dics_df   =   pd.DataFrame()

    for col in LABEL_COL:
        # Not NaN index
        idx                =  ~_df[col].isna()
        colx               =  col + '_code'
        _df.loc[idx, colx] =  le.fit(_df.loc[idx, col]).transform(_df.loc[idx, col])
        d_                 =  {l: i for i, l in enumerate(le.classes_)}
        d_df               =   pd.DataFrame({'Column' : col, 'dict' : d_})
        dics_df            =   pd.concat([dics_df, d_df])
    print("\nCols {}".format(_df.columns))
    return (_df, dics_df)

In [23]:
def code_encode(df):
    from    sklearn    import  preprocessing 
    to_encode   =  df.select_dtypes(exclude=["number","bool"]).columns.tolist()
    le          =  preprocessing.LabelEncoder()
    for col in to_encode:
        le.fit(df[col])        
        le_name_mapping =   dict(zip(le.classes_, le.transform(le.classes_)))
        df[col]         =   np.round(le.fit_transform(df[col]),0)
        if col == 'payer_code':
           df[col] = df[col].replace(17, np.NaN) 
        elif col == 'race':
           df[col] = df[col].replace(5, np.NaN)        
        elif col == 'medical_specialty':
           df[col] = df[col].replace(72, np.NaN)
        elif col == 'diag_1':
           df[col] = df[col].replace(716, np.NaN)
        elif col == 'diag_2':
           df[col] = df[col].replace(748, np.NaN)            
        elif col == 'diag_3':
           df[col] = df[col].replace(789, np.NaN)  
            
        df[col]         =   df[col].astype( 'category' ) 
        print("\n Col: {0} Code :{1}".format(col, le_name_mapping))


In [24]:
import  os

folder =  r'G:\DSE-FT-B-JUN24-G2'
pd.set_option('display.max_columns', None)
os.chdir(folder)

### 2) Read data

In [26]:
file   =  './Data/diabetic_data.csv'
df     =  pd.read_csv(file, keep_default_na=False, na_values=['?'])

In [27]:
df.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [28]:
df['A1Cresult'].value_counts()

A1Cresult
None    84748
>8       8216
Norm     4990
>7       3812
Name: count, dtype: int64

In [29]:
df.head(10).T

0          1  \
encounter_id                               2278392     149190   
patient_nbr                                8222157   55629189   
race                                     Caucasian  Caucasian   
gender                                      Female     Female   
age                                         [0-10)    [10-20)   
weight                                         NaN        NaN   
admission_type_id                                6          1   
discharge_disposition_id                        25          1   
admission_source_id                              1          7   
time_in_hospital                                 1          3   
payer_code                                     NaN        NaN   
medical_specialty         Pediatrics-Endocrinology        NaN   
num_lab_procedures                              41         59   
num_procedures                                   0          0   
num_medications                                  1         18   
number_outpatient                                0          0   
number_emergency                                 0          0   
number_inpatient                                 0          0   
diag_1                                      250.83        276   
diag_2                                         NaN     250.01   
diag_3                                         NaN        255   
number_diagnoses                                 1          9   
max_glu_serum                                 None       None   
A1Cresult                                     None       None   
metformin                                       No         No   
repaglinide                                     No         No   
nateglinide                                     No         No   
chlorpropamide                                  No         No   
glimepiride                                     No         No   
acetohexamide                                   No         No   
glipizide                                       No         No   
glyburide                                       No         No   
tolbutamide                                     No         No   
pioglitazone                                    No         No   
rosiglitazone                                   No         No   
acarbose                                        No         No   
miglitol                                        No         No   
troglitazone                                    No         No   
tolazamide                                      No         No   
examide                                         No         No   
citoglipton                                     No         No   
insulin                                         No         Up   
glyburide-metformin                             No         No   
glipizide-metformin                             No         No   
glimepiride-pioglitazone                        No         No   
metformin-rosiglitazone                         No         No   
metformin-pioglitazone                          No         No   
change                                          No         Ch   
diabetesMed                                     No        Yes   
readmitted                                      NO        >30   

                                        2          3          4          5  \
encounter_id                        64410     500364      16680      35754   
patient_nbr                      86047875   82442376   42519267   82637451   
race                      AfricanAmerican  Caucasian  Caucasian  Caucasian   
gender                             Female       Male       Male       Male   
age                               [20-30)    [30-40)    [40-50)    [50-60)   
weight                                NaN        NaN        NaN        NaN   
admission_type_id                       1          1          1          2   
discharge_disposition_id                1          1          1          1   
admission_source_id            

### 3) Understanding Dataset and Domain

### Variable categorization (count of numeric and categorical)

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      99493 non-null   object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    3197 non-null    object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                61510 non-null   object
 11  medical_specialty         51817 non-null   object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

In [33]:
df.select_dtypes(include = 'object').head()

race  gender      age weight payer_code  \
0        Caucasian  Female   [0-10)    NaN        NaN   
1        Caucasian  Female  [10-20)    NaN        NaN   
2  AfricanAmerican  Female  [20-30)    NaN        NaN   
3        Caucasian    Male  [30-40)    NaN        NaN   
4        Caucasian    Male  [40-50)    NaN        NaN   

          medical_specialty  diag_1  diag_2 diag_3 max_glu_serum A1Cresult  \
0  Pediatrics-Endocrinology  250.83     NaN    NaN          None      None   
1                       NaN     276  250.01    255          None      None   
2                       NaN     648     250    V27          None      None   
3                       NaN       8  250.43    403          None      None   
4                       NaN     197     157    250          None      None   

  metformin repaglinide nateglinide chlorpropamide glimepiride acetohexamide  \
0        No          No          No             No          No            No   
1        No          No          No             No          No            No   
2        No          No          No             No          No            No   
3        No          No          No             No          No            No   
4        No          No          No             No          No            No   

  glipizide glyburide tolbutamide pioglitazone rosiglitazone acarbose  \
0        No        No          No           No            No       No   
1        No        No          No           No            No       No   
2    Steady        No          No           No            No       No   
3        No        No          No           No            No       No   
4    Steady        No          No           No            No       No   

  miglitol troglitazone tolazamide examide citoglipton insulin  \
0       No           No         No      No          No      No   
1       No           No         No      No          No      Up   
2       No           No         No      No          No      No   
3       No           No         No      No          No      Up   
4       No           No         No      No          No  Steady   

  glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
0                  No                  No                       No   
1                  No                  No                       No   
2                  No                  No                       No   
3                  No                  No                       No   
4                  No                  No                       No   

  metformin-rosiglitazone metformin-pioglitazone change diabetesMed readmitted  
0                      No                     No     No          No         NO  
1                      No                     No     Ch         Yes        >30  
2                      No                     No     No         Yes         NO  
3                      No                     No     Ch         Yes         NO  
4                      No                     No     Ch         Yes         NO

In [34]:
df.select_dtypes(include = 'object').columns

Index(['race', 'gender', 'age', 'weight', 'payer_code', 'medical_specialty',
       'diag_1', 'diag_2', 'diag_3', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [35]:
df.select_dtypes(include = 'number').head()

encounter_id  patient_nbr  admission_type_id  discharge_disposition_id  \
0       2278392      8222157                  6                        25   
1        149190     55629189                  1                         1   
2         64410     86047875                  1                         1   
3        500364     82442376                  1                         1   
4         16680     42519267                  1                         1   

   admission_source_id  time_in_hospital  num_lab_procedures  num_procedures  \
0                    1                 1                  41               0   
1                    7                 3                  59               0   
2                    7                 2                  11               5   
3                    7                 2                  44               1   
4                    7                 1                  51               0   

   num_medications  number_outpatient  number_emergency  number_inpatient  \
0                1                  0                 0                 0   
1               18                  0                 0                 0   
2               13                  2                 0                 1   
3               16                  0                 0                 0   
4                8                  0                 0                 0   

   number_diagnoses  
0                 1  
1                 9  
2                 6  
3                 7  
4                 5

In [36]:
df.select_dtypes(include = 'number').columns

Index(['encounter_id', 'patient_nbr', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient',
       'number_diagnoses'],
      dtype='object')

Inferences:
    
    -->Number of Catagorical Columns : 37
    
    -->Number of Numerical Columns : 13 of which one is our target variable

### c) Pre Processing Data Analysis (count of missing/ null values, redundant columns, etc.)

<b>Null or Missing Values</b>

In [40]:
missing_zero_values_table(df)

Your selected dataframe has 50 columns and 101766 Rows.
There are 7 columns that have missing values.


Count of Missing Values  % Missing Values Data Type
weight                               98569             96.86    object
medical_specialty                    49949             49.08    object
payer_code                           40256             39.56    object
race                                  2273              2.23    object
diag_3                                1423              1.40    object
diag_2                                 358              0.35    object
diag_1                                  21              0.02    object

### Observations

We observe that the column, weight is having 96.86% missing values. **If you have a column with more than 80% missing values, then it is better to drop the column.**
So, we need to drop the column, weight from our dataset.

In [42]:
drop_cols =  ['weight']
df.drop(drop_cols, inplace = True, axis = 1)

### Identify unwanted columns

ID variables do not add value to our analysis and we need to drop them from the dataset.
There are two ID variables, namely: 
* 1) encounter_id
* 2) patient_nbr

In [45]:
drop_cols =  ['encounter_id', 'patient_nbr']
df.drop(drop_cols, inplace = True, axis = 1)

### Check the Target variable

We shall derive the target variable, 'Target' as follows:
* Target value: 0 -- Readmitted NO and Readmitted > 30
* Target value: 1 -- Readmitted < 30

In [48]:
# create a list of our conditions
conditions = [
        (df['readmitted'] ==  '>30') | (df['readmitted'] == 'NO'),    
        (df['readmitted'] == '<30' )
]
# create a list of the values we want to assign for each condition
values = [0, 1]

# create a new column and use np.select to assign values to it using our lists as arguments
df['Target'] = np.select(conditions, values)

In [49]:
df.columns

Index(['race', 'gender', 'age', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'Target'],
      dtype='object')

### Check for constant value columns

In [51]:
drop_cols = chk_constant_rows(df)
print(drop_cols)


Unique Column: examide Count 1

Unique Column: citoglipton Count 1
['examide', 'citoglipton']


### Observation

We need to remove the columns with constant values from our dataset.

We have constant values in the two columns, namely:
##### 1) examide
##### 2) citoglipton

In [53]:
df.drop(drop_cols, inplace = True, axis = 1)

### Check for duplicated rows

In [55]:
df.duplicated(keep = 'first').sum()

0

### Observation

There are no duplicates in our dataset.

In [57]:
drop_cols= 'readmitted' ## We remove the old variable, readmitted which is used to derive our target variable
df.drop(drop_cols, inplace = True, axis = 1)

In [58]:
df.shape

(101766, 45)

In [59]:
df.columns

Index(['race', 'gender', 'age', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'Target'],
      dtype='object')

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 45 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   race                      99493 non-null   object
 1   gender                    101766 non-null  object
 2   age                       101766 non-null  object
 3   admission_type_id         101766 non-null  int64 
 4   discharge_disposition_id  101766 non-null  int64 
 5   admission_source_id       101766 non-null  int64 
 6   time_in_hospital          101766 non-null  int64 
 7   payer_code                61510 non-null   object
 8   medical_specialty         51817 non-null   object
 9   num_lab_procedures        101766 non-null  int64 
 10  num_procedures            101766 non-null  int64 
 11  num_medications           101766 non-null  int64 
 12  number_outpatient         101766 non-null  int64 
 13  number_emergency          101766 non-null  int64 
 14  numb

### Observations

We have **45 variables left** after dropping unwanted columns such as ID variables, constant value columns etc.

There are 12 numeric columns, 27 categorical columns without null values and 6 categorical columns with null values.

### We shall proceed to treat missing values.


In [62]:
n12Var = ['admission_type_id','discharge_disposition_id', 'admission_source_id', 'time_in_hospital','num_lab_procedures','num_procedures',\
'num_medications','number_outpatient','number_emergency', 'number_inpatient', 'number_diagnoses', 'Target']

cat27_notnull = ['gender', 'age', 'max_glu_serum', 'A1Cresult', 'metformin','repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride','acetohexamide',\
'glipizide', 'glyburide', 'tolbutamide','pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone','tolazamide', 'insulin', 'glyburide-metformin',\
'glipizide-metformin','glimepiride-pioglitazone', 'metformin-rosiglitazone','metformin-pioglitazone', 'change', 'diabetesMed']


cat6_null =   ['race', 'payer_code', 'medical_specialty','diag_1', 'diag_2', 'diag_3']

### How to treat missing values?

### Methods of treating missing values:
1) Remove Rows With Missing Values: Remove rows that contain missing values.
2) Impute Missing Values: Replace missing values with sensible values.
3) Impute Missing Values with KNN Imputer: Impute missing values using K nearest neighbors.
4) Impute Missing Values with Iterative Imputer: Impute missing values in multiple features using iterative imputation.
IterativeImputer in SKlearn. It is a strategy for imputing missing values by modeling each feature with missing values as a function of other features in a round-robin fashion.

###  IterativeImputer() from sklearn

A more sophisticated approach is to use the IterativeImputer class, which models each feature with missing values as a function of other features, and uses that estimate for imputation. It does so in an iterated round-robin fashion: at each step, a feature column is designated as output y and the other feature columns are treated as inputs X. A regressor is fit on (X, y) for known y. Then, the regressor is used to predict the missing values of y. This is done for each feature in an iterative fashion, and then is repeated for max_iter imputation rounds. The results of the final imputation round are returned.

In [65]:
cols_notna =  cat27_notnull + n12Var
print(cols_notna)

['gender', 'age', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'Target']


In [66]:
Nonnull_Cols_df  =   df.loc[:, cols_notna]

## Convert categorical columns to numeric columns to facilitate imputing and building models

In [68]:
enc_df                   =    df.copy()
cat_cols                 =    cat6_null +  cat27_notnull

In [69]:
print("\nBefore Label Encoding")
print(enc_df.head(3).T)


Before Label Encoding
                                                 0          1                2
race                                     Caucasian  Caucasian  AfricanAmerican
gender                                      Female     Female           Female
age                                         [0-10)    [10-20)          [20-30)
admission_type_id                                6          1                1
discharge_disposition_id                        25          1                1
admission_source_id                              1          7                7
time_in_hospital                                 1          3                2
payer_code                                     NaN        NaN              NaN
medical_specialty         Pediatrics-Endocrinology        NaN              NaN
num_lab_procedures                              41         59               11
num_procedures                                   0          0                5
num_medications              

In [70]:
dic_df               =  pd.DataFrame()
LABEL_COL            =  cat_cols 
transformed, dic_df  =  label(enc_df, LABEL_COL)


Cols Index(['race', 'gender', 'age', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'Target',
       'race_code', 'payer_code_code', 'medical_specialty_code', 'diag_1_code',
       'diag_2_code', 'diag_3_code', 'gender_code', 'age_code',
       'max_glu_seru

In [71]:
transformed.head(2).T

0          1
race                           Caucasian  Caucasian
gender                            Female     Female
age                               [0-10)    [10-20)
admission_type_id                      6          1
discharge_disposition_id              25          1
...                                  ...        ...
glimepiride-pioglitazone_code        0.0        0.0
metformin-rosiglitazone_code         0.0        0.0
metformin-pioglitazone_code          0.0        0.0
change_code                          1.0        0.0
diabetesMed_code                     0.0        1.0

[78 rows x 2 columns]

In [72]:
transformed.columns

Index(['race', 'gender', 'age', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'Target',
       'race_code', 'payer_code_code', 'medical_specialty_code', 'diag_1_code',
       'diag_2_code', 'diag_3_code', 'gender_code', 'age_code',
       'max_glu_serum_code

In [73]:
enc27_notnull = ['gender_code', 'age_code', 'max_glu_serum_code', 'A1Cresult_code', 'metformin_code','repaglinide_code', 'nateglinide_code',\
                 'chlorpropamide_code', 'glimepiride_code','acetohexamide_code','glipizide_code', 'glyburide_code', 'tolbutamide_code',\
                 'pioglitazone_code', 'rosiglitazone_code', 'acarbose_code', 'miglitol_code', 'troglitazone_code','tolazamide_code', 'insulin_code',\
                 'glyburide-metformin_code','glipizide-metformin_code','glimepiride-pioglitazone_code', 'metformin-rosiglitazone_code',\
                 'metformin-pioglitazone_code', 'change_code', 'diabetesMed_code']
enc6_null    =   ['race_code', 'payer_code_code', 'medical_specialty_code','diag_1_code', 'diag_2_code', 'diag_3_code']

# impute missing value

# Initialize IterativeImputer with DecisionTreeRegressor

In [75]:
imp_cols = n12Var  + enc27_notnull + enc6_null

In [76]:
len(imp_cols)

45

In [77]:
transformed.columns

Index(['race', 'gender', 'age', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3',
       'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'Target',
       'race_code', 'payer_code_code', 'medical_specialty_code', 'diag_1_code',
       'diag_2_code', 'diag_3_code', 'gender_code', 'age_code',
       'max_glu_serum_code

In [78]:
len(transformed.columns)

78

In [79]:
imputed_encoded_data   =   transformed.loc[:, imp_cols]

In [80]:
imputed_encoded_data.shape

(101766, 45)

In [81]:
imputer         = IterativeImputer(estimator = DecisionTreeRegressor(), max_iter = 10, random_state = 0)
imputed_data    = imputer.fit_transform(imputed_encoded_data)

In [82]:
imputed_df      = pd.DataFrame(imputed_data, columns = imputed_encoded_data.columns)

In [83]:
print("Imputed Dataset:")
print(imputed_df.head(12))

Imputed Dataset:
    admission_type_id  discharge_disposition_id  admission_source_id  \
0                 6.0                      25.0                  1.0   
1                 1.0                       1.0                  7.0   
2                 1.0                       1.0                  7.0   
3                 1.0                       1.0                  7.0   
4                 1.0                       1.0                  7.0   
5                 2.0                       1.0                  2.0   
6                 3.0                       1.0                  2.0   
7                 1.0                       1.0                  7.0   
8                 2.0                       1.0                  4.0   
9                 3.0                       3.0                  4.0   
10                1.0                       1.0                  7.0   
11                2.0                       1.0                  4.0   

    time_in_hospital  num_lab_procedures  num_

In [84]:
missing_zero_values_table(imputed_df)

Your selected dataframe has 45 columns and 101766 Rows.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Count of Missing Values, % Missing Values, Data Type]
Index: []

### Observations

Now we have cleaned data that can be used for model building.
We shall use the dataframe, **imputed_df** for model buidling and all the features are label encoded.

In [86]:
csv_file1 = './Data/DSEFTB-JUN24-G2-Models_' + str(datetime.now().strftime('%Y_%m_%d_%H_%M_%S')) + '.csv'
imputed_df.to_csv(csv_file1, index = False)

### Data file for performing EDA

For the purpose of EDA, we shall use the original values for categorical variables without null values.
For the categorical variables, with null values, we shall get the label name by reversing key and value pair in the dictionary.

### We shall copy the columns with orginal values from not null columns first

In [89]:
cat27_notnull = ['gender', 'age', 'max_glu_serum', 'A1Cresult', 'metformin','repaglinide', 'nateglinide',\
                 'chlorpropamide', 'glimepiride','acetohexamide','glipizide', 'glyburide', 'tolbutamide',\
                 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone','tolazamide', 'insulin',\
                 'glyburide-metformin','glipizide-metformin','glimepiride-pioglitazone', 'metformin-rosiglitazone',\
                 'metformin-pioglitazone', 'change', 'diabetesMed']
EDA_cols       =   n12Var + cat27_notnull

In [90]:
df_EDA            =  df.loc[:, EDA_cols]

In [91]:
df_EDA.shape

(101766, 39)

### We shall decode the encoded values for the null value columns after imputation

In [93]:
print(dic_df.shape)

(2441, 2)


In [94]:
print("\nMapping")
print(dic_df)


Mapping
                                 Column  dict
AfricanAmerican                    race     0
Asian                              race     1
Caucasian                          race     2
Hispanic                           race     3
Other                              race     4
...                                 ...   ...
Steady           metformin-pioglitazone     1
Ch                               change     0
No                               change     1
No                          diabetesMed     0
Yes                         diabetesMed     1

[2441 rows x 2 columns]


In [95]:
category_cols         =    enc6_null + enc27_notnull
print(category_cols)

['race_code', 'payer_code_code', 'medical_specialty_code', 'diag_1_code', 'diag_2_code', 'diag_3_code', 'gender_code', 'age_code', 'max_glu_serum_code', 'A1Cresult_code', 'metformin_code', 'repaglinide_code', 'nateglinide_code', 'chlorpropamide_code', 'glimepiride_code', 'acetohexamide_code', 'glipizide_code', 'glyburide_code', 'tolbutamide_code', 'pioglitazone_code', 'rosiglitazone_code', 'acarbose_code', 'miglitol_code', 'troglitazone_code', 'tolazamide_code', 'insulin_code', 'glyburide-metformin_code', 'glipizide-metformin_code', 'glimepiride-pioglitazone_code', 'metformin-rosiglitazone_code', 'metformin-pioglitazone_code', 'change_code', 'diabetesMed_code']


In [157]:
dic_df.head().T

AfricanAmerican Asian Caucasian Hispanic Other
Column            race  race      race     race  race
dict                 0     1         2        3     4

In [ ]:
rev_dicts         =    []
dic_df
for i in range(len(category_cols)):
    x_        =  dicts_l[i]
    colx      =  category_cols[i]
    ## print("\nCol : {0}-- dict {1}".format(colx, x_))
    rev_dict  =  dict(map(lambda key: (x_[key], key), x_.keys()))
    ## print("\nReverse encoding for col {0} \n {1}".format(colx, rev_dict))    
    rev_dicts.append(dict(rev_dict))

In [ ]:
rev_dicts

In [ ]:
imputed_na_df   =   imputed_df.loc[:, enc6_null]

In [ ]:
df_EDA.shape

In [ ]:
imputed_na_df.shape

In [ ]:
dicts             =   rev_dicts
for  i  in  range(len(enc6_null)):
    col       =   enc6_null[i]
    dic       =   dicts[i]
    print("\n {}".format(col))
    print("\n {}".format(dic))
    imputed_na_df.replace({col: dic}, inplace = True)

df_c          =   pd.concat([df_EDA, imputed_na_df],axis = 1, ignore_index=True)
print(df_c.shape)

In [ ]:
df_c_columns    =   category_cols + n12Var
df_c.columns    =   df_c_columns
print(df_c)

### WE shall construct the dataframe for doing EDA

In [ ]:
 df.select_dtypes(include =['int64','float64']) 
num_cols_EDA      =  df_EDA.columns
print(num_cols_EDA)

In [ ]:
EDA_cols_notimputed  =   ['admission_type_id', 'discharge_disposition_id', 'admission_source_id','time_in_hospital', 'num_lab_procedures',\
                          'num_procedures','num_medications', 'number_outpatient', 'number_emergency','number_inpatient', 'number_diagnoses',\
                          'gender', 'age', 'max_glu_serum', 'A1Cresult','metformin','repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',\
                          'acetohexamide','glipizide', 'glyburide', 'tolbutamide','pioglitazone','rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',\
                          'tolazamide', 'insulin', 'glyburide-metformin', 'glipizide-metformin','glimepiride-pioglitazone','metformin-rosiglitazone',\
                          'metformin-pioglitazone', 'change','diabetesMed', 'Target'] 
print(EDA_cols_notimputed)

In [ ]:
print(len(EDA_cols_notimputed))

In [ ]:
['race','payer_code','medical_specialty','diag_1','diag_2','diag_3']

In [ ]:
dicts             =   [d2_age, d2_gender]
for  i  in  range(len(Null_cols)):
    col       =   Null_cols[i]
    dic       =   dicts[i]
    print("\n {}".format(col))
    print("\n {}".format(dic))
    imputed_df.replace({col: dic}, inplace = True)

df_c  = pd.concat([df, imputed_df],axis = 1, ignore_index=True)
df_c.columns =  ['ID', 'Age', 'Salary', 'Gender', 'ID_R', 'Age_R', 'Salary_R', 'Gender_R']
print(df_c)

In [ ]:
transformed.head().T

In [ ]:
lreg         =  LinearRegression()
imp          =  IterativeImputer(estimator= lreg, missing_values = np.nan, max_iter = 10, verbose = 2,\
                                 imputation_order= 'roman',random_state = 0)
X            =  np.round(imp.fit_transform(enc_df))

In [ ]:
imputed_df    =    pd.DataFrame(X, columns = enc_df.columns)
imputed_df    =    imputed_df.astype(int)
df_for_model  =    imputed_df.copy()

In [ ]:
csv_file1 = './Data/DSEFTB-JUN24-G2-Models_' + str(datetime.now().strftime('%Y_%m_%d_%H_%M_%S')) + '.csv'
df_for_model.to_csv(csv_file1, index = False)

In [ ]:
df_for_model.head(10).T

In [ ]:
def rev_dict(d1):
    return dict(map(lambda key: (d1[key], key), d1.keys()))

In [ ]:
dict_race               =   {'AfricanAmerican': 0, 'Asian': 1, 'Caucasian': 2, 'Hispanic': 3, 'Other': 4}
d2_race                 =   rev_dict(dict_race)
print(d2_race)

'race','payer_code','medical_specialty','diag_1','diag_2','diag_3'

In [ ]:
dict_payer_code         =   {'BC': 0, 'CH': 1, 'CM': 2, 'CP': 3, 'DM': 4, 'FR': 5, 'HM': 6, 'MC': 7, 'MD': 8, 'MP': 9, 'OG': 10,\
                             'OT': 11, 'PO': 12, 'SI': 13, 'SP': 14, 'UN': 15, 'WC': 16}
d2_payer_code           =   rev_dict(dict_payer_code)
print(d2_payer_code)

In [ ]:
dict_medical_specialty  =   {'AllergyandImmunology': 0, 'Anesthesiology': 1, 'Anesthesiology-Pediatric': 2, 'Cardiology': 3,\
                             'Cardiology-Pediatric': 4, 'DCPTEAM': 5, 'Dentistry': 6, 'Dermatology': 7, 'Emergency/Trauma': 8,\
                             'Endocrinology': 9, 'Endocrinology-Metabolism': 10, 'Family/GeneralPractice': 11, 'Gastroenterology': 12,\
                             'Gynecology': 13, 'Hematology': 14, 'Hematology/Oncology': 15, 'Hospitalist': 16, 'InfectiousDiseases': 17,\
                             'InternalMedicine': 18, 'Nephrology': 19, 'Neurology': 20, 'Neurophysiology': 21,\
                             'Obsterics&Gynecology-GynecologicOnco': 22, 'Obstetrics': 23, 'ObstetricsandGynecology': 24,\
                             'Oncology': 25, 'Ophthalmology': 26, 'Orthopedics': 27, 'Orthopedics-Reconstructive': 28,\
                             'Osteopath': 29, 'Otolaryngology': 30, 'OutreachServices': 31, 'Pathology': 32, 'Pediatrics': 33,\
                             'Pediatrics-AllergyandImmunology': 34, 'Pediatrics-CriticalCare': 35, 'Pediatrics-EmergencyMedicine': 36,\
                             'Pediatrics-Endocrinology': 37, 'Pediatrics-Hematology-Oncology': 38, 'Pediatrics-InfectiousDiseases': 39,\
                             'Pediatrics-Neurology': 40, 'Pediatrics-Pulmonology': 41, 'Perinatology': 42, 'PhysicalMedicineandRehabilitation': 43,\
                             'PhysicianNotFound': 44, 'Podiatry': 45, 'Proctology': 46, 'Psychiatry': 47, 'Psychiatry-Addictive': 48,\
                             'Psychiatry-Child/Adolescent': 49, 'Psychology': 50, 'Pulmonology': 51, 'Radiologist': 52, 'Radiology': 53,\
                             'Resident': 54, 'Rheumatology': 55, 'Speech': 56, 'SportsMedicine': 57, 'Surgeon': 58, 'Surgery-Cardiovascular': 59,\
                             'Surgery-Cardiovascular/Thoracic': 60, 'Surgery-Colon&Rectal': 61, 'Surgery-General': 62, 'Surgery-Maxillofacial': 63,\
                             'Surgery-Neuro': 64, 'Surgery-Pediatric': 65, 'Surgery-Plastic': 66, 'Surgery-PlasticwithinHeadandNeck': 67,\
                             'Surgery-Thoracic': 68, 'Surgery-Vascular': 69, 'SurgicalSpecialty': 70, 'Urology': 71}

d2_medical_specialty           =   rev_dict(dict_medical_specialty)
print(d2_medical_specialty)

In [ ]:
dict_diag_1             =    {'10': 0, '11': 1, '110': 2, '112': 3, '114': 4, '115': 5, '117': 6, '131': 7, '133': 8, '135': 9, '136': 10, '141': 11,\
                              '142': 12, '143': 13, '145': 14, '146': 15, '147': 16, '148': 17, '149': 18, '150': 19, '151': 20, '152': 21, '153': 22,\
                              '154': 23, '155': 24, '156': 25, '157': 26, '158': 27, '160': 28, '161': 29, '162': 30, '163': 31, '164': 32, '170': 33,\
                              '171': 34, '172': 35, '173': 36, '174': 37, '175': 38, '179': 39, '180': 40, '182': 41, '183': 42, '184': 43, '185': 44,\
                              '187': 45, '188': 46, '189': 47, '191': 48, '192': 49, '193': 50, '194': 51, '195': 52, '196': 53, '197': 54, '198': 55,\
                              '199': 56, '200': 57, '201': 58, '202': 59, '203': 60, '204': 61, '205': 62, '207': 63, '208': 64, '210': 65, '211': 66,\
                              '212': 67, '214': 68, '215': 69, '216': 70, '217': 71, '218': 72, '219': 73, '220': 74, '223': 75, '225': 76, '226': 77,\
                              '227': 78, '228': 79, '229': 80, '23': 81, '230': 82, '233': 83, '235': 84, '236': 85, '237': 86, '238': 87, '239': 88,\
                              '240': 89, '241': 90, '242': 91, '244': 92, '245': 93, '246': 94, '250': 95, '250.01': 96, '250.02': 97, '250.03': 98,\
                              '250.1': 99, '250.11': 100, '250.12': 101, '250.13': 102, '250.2': 103, '250.21': 104, '250.22': 105, '250.23': 106,\
                              '250.3': 107, '250.31': 108, '250.32': 109, '250.33': 110, '250.4': 111, '250.41': 112, '250.42': 113, '250.43': 114,\
                              '250.5': 115, '250.51': 116, '250.52': 117, '250.53': 118, '250.6': 119, '250.7': 120, '250.8': 121, '250.81': 122,\
                              '250.82': 123, '250.83': 124, '250.9': 125, '250.91': 126, '250.92': 127, '250.93': 128, '251': 129, '252': 130,\
                              '253': 131, '255': 132, '261': 133, '262': 134, '263': 135, '266': 136, '27': 137, '271': 138, '272': 139, '273': 140,\
                              '274': 141, '275': 142, '276': 143, '277': 144, '278': 145, '279': 146, '280': 147, '281': 148, '282': 149, '283': 150,\
                              '284': 151, '285': 152, '286': 153, '287': 154, '288': 155, '289': 156, '290': 157, '291': 158, '292': 159, '293': 160,\
                              '294': 161, '295': 162, '296': 163, '297': 164, '298': 165, '299': 166, '3': 167, '300': 168, '301': 169, '303': 170,\
                              '304': 171, '305': 172, '306': 173, '307': 174, '308': 175, '309': 176, '31': 177, '310': 178, '311': 179, '312': 180,\
                              '314': 181, '318': 182, '320': 183, '322': 184, '323': 185, '324': 186, '325': 187, '327': 188, '331': 189, '332': 190,\
                              '333': 191, '334': 192, '335': 193, '336': 194, '337': 195, '338': 196, '34': 197, '340': 198, '341': 199, '342': 200,\
                              '344': 201, '345': 202, '346': 203, '347': 204, '348': 205, '349': 206, '35': 207, '350': 208, '351': 209, '352': 210,\
                              '353': 211, '354': 212, '355': 213, '356': 214, '357': 215, '358': 216, '359': 217, '36': 218, '360': 219, '361': 220,\
                              '362': 221, '363': 222, '365': 223, '366': 224, '368': 225, '369': 226, '370': 227, '372': 228, '373': 229, '374': 230,\
                              '375': 231, '376': 232, '377': 233, '378': 234, '379': 235, '38': 236, '380': 237, '381': 238, '382': 239, '383': 240,\
                              '384': 241, '385': 242, '386': 243, '388': 244, '389': 245, '39': 246, '391': 247, '394': 248, '395': 249, '396': 250,\
                              '397': 251, '398': 252, '401': 253, '402': 254, '403': 255, '404': 256, '405': 257, '41': 258, '410': 259, '411': 260,\
                              '412': 261, '413': 262, '414': 263, '415': 264, '416': 265, '417': 266, '42': 267, '420': 268, '421': 269, '422': 270,\
                              '423': 271, '424': 272, '425': 273, '426': 274, '427': 275, '428': 276, '429': 277, '430': 278, '431': 279, '432': 280,\
                              '433': 281, '434': 282, '435': 283, '436': 284, '437': 285, '438': 286, '440': 287, '441': 288, '442': 289, '443': 290,\
                              '444': 291, '445': 292, '446': 293, '447': 294, '448': 295, '451': 296, '452': 297, '453': 298, '454': 299, '455': 300,\
                              '456': 301, '457': 302, '458': 303, '459': 304, '461': 305, '462': 306, '463': 307, '464': 308, '465': 309, '466': 310,\
                              '47': 311, '470': 312, '471': 313, '473': 314, '474': 315, '475': 316, '477': 317, '478': 318, '48': 319, '480': 320,\
                              '481': 321, '482': 322, '483': 323, '485': 324, '486': 325, '487': 326, '49': 327, '490': 328, '491': 329, '492': 330,\
                              '493': 331, '494': 332, '495': 333, '496': 334, '5': 335, '500': 336, '501': 337, '506': 338, '507': 339, '508': 340,\
                              '510': 341, '511': 342, '512': 343, '513': 344, '514': 345, '515': 346, '516': 347, '518': 348, '519': 349, '52': 350,\
                              '521': 351, '522': 352, '523': 353, '524': 354, '526': 355, '527': 356, '528': 357, '529': 358, '53': 359, '530': 360,\
                              '531': 361, '532': 362, '533': 363, '534': 364, '535': 365, '536': 366, '537': 367, '54': 368, '540': 369, '541': 370,\
                              '542': 371, '543': 372, '550': 373, '551': 374, '552': 375, '553': 376, '555': 377, '556': 378, '557': 379, '558': 380,\
                              '560': 381, '562': 382, '564': 383, '565': 384, '566': 385, '567': 386, '568': 387, '569': 388, '57': 389, '570': 390,\
                              '571': 391, '572': 392, '573': 393, '574': 394, '575': 395, '576': 396, '577': 397, '578': 398, '579': 399, '58': 400,\
                              '580': 401, '581': 402, '582': 403, '583': 404, '584': 405, '585': 406, '586': 407, '588': 408, '590': 409, '591': 410,\
                              '592': 411, '593': 412, '594': 413, '595': 414, '596': 415, '598': 416, '599': 417, '600': 418, '601': 419, '602': 420,\
                              '603': 421, '604': 422, '605': 423, '607': 424, '608': 425, '61': 426, '610': 427, '611': 428, '614': 429, '615': 430,\
                              '616': 431, '617': 432, '618': 433, '619': 434, '620': 435, '621': 436, '622': 437, '623': 438, '625': 439, '626': 440,\
                              '627': 441, '632': 442, '633': 443, '634': 444, '637': 445, '640': 446, '641': 447, '642': 448, '643': 449, '644': 450,\
                              '645': 451, '646': 452, '647': 453, '648': 454, '649': 455, '652': 456, '653': 457, '654': 458, '655': 459, '656': 460,\
                              '657': 461, '658': 462, '659': 463, '66': 464, '660': 465, '661': 466, '663': 467, '664': 468, '665': 469, '669': 470,\
                              '671': 471, '674': 472, '680': 473, '681': 474, '682': 475, '683': 476, '684': 477, '685': 478, '686': 479, '690': 480,\
                              '691': 481, '692': 482, '693': 483, '694': 484, '695': 485, '696': 486, '698': 487, '7': 488, '70': 489, '700': 490,\
                              '703': 491, '704': 492, '705': 493, '706': 494, '707': 495, '708': 496, '709': 497, '710': 498, '711': 499, '714': 500,\
                              '715': 501, '716': 502, '717': 503, '718': 504, '719': 505, '720': 506, '721': 507, '722': 508, '723': 509, '724': 510,\
                              '725': 511, '726': 512, '727': 513, '728': 514, '729': 515, '730': 516, '731': 517, '732': 518, '733': 519, '734': 520,\
                              '735': 521, '736': 522, '737': 523, '738': 524, '745': 525, '746': 526, '747': 527, '75': 528, '751': 529, '753': 530,\
                              '756': 531, '759': 532, '78': 533, '780': 534, '781': 535, '782': 536, '783': 537, '784': 538, '785': 539, '786': 540,\
                              '787': 541, '788': 542, '789': 543, '79': 544, '790': 545, '791': 546, '792': 547, '793': 548, '794': 549, '795': 550,\
                              '796': 551, '797': 552, '799': 553, '8': 554, '800': 555, '801': 556, '802': 557, '803': 558, '804': 559, '805': 560,\
                              '806': 561, '807': 562, '808': 563, '810': 564, '812': 565, '813': 566, '814': 567, '815': 568, '816': 569, '817': 570,\
                              '82': 571, '820': 572, '821': 573, '822': 574, '823': 575, '824': 576, '825': 577, '826': 578, '827': 579, '831': 580,\
                              '832': 581, '833': 582, '834': 583, '835': 584, '836': 585, '837': 586, '838': 587, '839': 588, '84': 589, '840': 590,\
                              '842': 591, '843': 592, '844': 593, '845': 594, '846': 595, '847': 596, '848': 597, '850': 598, '851': 599, '852': 600,\
                              '853': 601, '854': 602, '860': 603, '861': 604, '862': 605, '863': 606, '864': 607, '865': 608, '866': 609, '867': 610,\
                              '868': 611, '870': 612, '871': 613, '873': 614, '875': 615, '878': 616, '879': 617, '88': 618, '880': 619, '881': 620,\
                              '882': 621, '883': 622, '885': 623, '886': 624, '890': 625, '891': 626, '892': 627, '893': 628, '895': 629, '897': 630,\
                              '9': 631, '903': 632, '904': 633, '906': 634, '911': 635, '913': 636, '914': 637, '915': 638, '916': 639, '917': 640,\
                              '919': 641, '920': 642, '921': 643, '922': 644, '923': 645, '924': 646, '928': 647, '933': 648, '934': 649, '935': 650,\
                              '936': 651, '939': 652, '94': 653, '941': 654, '942': 655, '944': 656, '945': 657, '952': 658, '955': 659, '957': 660,\
                              '958': 661, '959': 662, '962': 663, '963': 664, '964': 665, '965': 666, '966': 667, '967': 668, '968': 669, '969': 670,\
                              '97': 671, '970': 672, '971': 673, '972': 674, '973': 675, '974': 676, '975': 677, '976': 678, '977': 679, '98': 680,\
                              '980': 681, '982': 682, '983': 683, '986': 684, '987': 685, '988': 686, '989': 687, '990': 688, '991': 689, '992': 690,\
                              '994': 691, '995': 692, '996': 693, '997': 694, '998': 695, '999': 696, 'E909': 697, 'V07': 698, 'V25': 699, 'V26': 700,\
                              'V43': 701, 'V45': 702, 'V51': 703, 'V53': 704, 'V54': 705, 'V55': 706, 'V56': 707, 'V57': 708, 'V58': 709, 'V60': 710,\
                              'V63': 711, 'V66': 712, 'V67': 713, 'V70': 714, 'V71': 715}

d2_diag_1           =   rev_dict(dict_diag_1)
print(d2_diag_1)

In [ ]:
dict_diag_2             =    {'11': 0, '110': 1, '111': 2, '112': 3, '114': 4, '115': 5, '117': 6, '123': 7, '130': 8, '131': 9, '135': 10, '136': 11, '137': 12, '138': 13, '140': 14, '141': 15, '145': 16, '150': 17, '151': 18, '152': 19, '153': 20, '154': 21, '155': 22, '156': 23, '157': 24, '162': 25, '163': 26, '164': 27, '171': 28, '172': 29, '173': 30, '174': 31, '179': 32, '180': 33, '182': 34, '183': 35, '185': 36, '186': 37, '188': 38, '189': 39, '191': 40, '192': 41, '193': 42, '195': 43, '196': 44, '197': 45, '198': 46, '199': 47, '200': 48, '201': 49, '202': 50, '203': 51, '204': 52, '205': 53, '208': 54, '211': 55, '212': 56, '214': 57, '215': 58, '217': 59, '218': 60, '220': 61, '223': 62, '225': 63, '226': 64, '227': 65, '228': 66, '232': 67, '233': 68, '235': 69, '238': 70, '239': 71, '240': 72, '241': 73, '242': 74, '244': 75, '245': 76, '246': 77, '250': 78, '250.01': 79, '250.02': 80, '250.03': 81, '250.1': 82, '250.11': 83, '250.12': 84, '250.13': 85, '250.2': 86, '250.21': 87, '250.22': 88, '250.23': 89, '250.3': 90, '250.31': 91, '250.32': 92, '250.33': 93, '250.4': 94, '250.41': 95, '250.42': 96, '250.43': 97, '250.5': 98, '250.51': 99, '250.52': 100, '250.53': 101, '250.6': 102, '250.7': 103, '250.8': 104, '250.81': 105, '250.82': 106, '250.83': 107, '250.9': 108, '250.91': 109, '250.92': 110, '250.93': 111, '251': 112, '252': 113, '253': 114, '255': 115, '256': 116, '258': 117, '259': 118, '260': 119, '261': 120, '262': 121, '263': 122, '266': 123, '268': 124, '269': 125, '27': 126, '270': 127, '271': 128, '272': 129, '273': 130, '274': 131, '275': 132, '276': 133, '277': 134, '278': 135, '279': 136, '280': 137, '281': 138, '282': 139, '283': 140, '284': 141, '285': 142, '286': 143, '287': 144, '288': 145, '289': 146, '290': 147, '291': 148, '292': 149, '293': 150, '294': 151, '295': 152, '296': 153, '297': 154, '298': 155, '299': 156, '300': 157, '301': 158, '302': 159, '303': 160, '304': 161, '305': 162, '306': 163, '307': 164, '308': 165, '309': 166, '31': 167, '310': 168, '311': 169, '312': 170, '314': 171, '316': 172, '317': 173, '318': 174, '319': 175, '320': 176, '322': 177, '323': 178, '324': 179, '325': 180, '327': 181, '331': 182, '332': 183, '333': 184, '335': 185, '336': 186, '337': 187, '338': 188, '34': 189, '340': 190, '341': 191, '342': 192, '343': 193, '344': 194, '345': 195, '346': 196, '347': 197, '348': 198, '349': 199, '35': 200, '350': 201, '351': 202, '352': 203, '353': 204, '354': 205, '355': 206, '356': 207, '357': 208, '358': 209, '359': 210, '360': 211, '362': 212, '364': 213, '365': 214, '366': 215, '368': 216, '369': 217, '372': 218, '373': 219, '374': 220, '376': 221, '377': 222, '378': 223, '379': 224, '38': 225, '380': 226, '381': 227, '382': 228, '383': 229, '386': 230, '388': 231, '389': 232, '394': 233, '395': 234, '396': 235, '397': 236, '398': 237, '40': 238, '401': 239, '402': 240, '403': 241, '404': 242, '405': 243, '41': 244, '410': 245, '411': 246, '412': 247, '413': 248, '414': 249, '415': 250, '416': 251, '42': 252, '420': 253, '421': 254, '422': 255, '423': 256, '424': 257, '425': 258, '426': 259, '427': 260, '428': 261, '429': 262, '430': 263, '431': 264, '432': 265, '433': 266, '434': 267, '435': 268, '436': 269, '437': 270, '438': 271, '440': 272, '441': 273, '442': 274, '443': 275, '444': 276, '446': 277, '447': 278, '448': 279, '451': 280, '452': 281, '453': 282, '454': 283, '455': 284, '456': 285, '457': 286, '458': 287, '459': 288, '46': 289, '460': 290, '461': 291, '462': 292, '463': 293, '464': 294, '465': 295, '466': 296, '470': 297, '472': 298, '473': 299, '474': 300, '475': 301, '477': 302, '478': 303, '480': 304, '481': 305, '482': 306, '483': 307, '484': 308, '485': 309, '486': 310, '487': 311, '490': 312, '491': 313, '492': 314, '493': 315, '494': 316, '495': 317, '496': 318, '5': 319, '500': 320, '501': 321, '506': 322, '507': 323, '508': 324, '510': 325, '511': 326, '512': 327, '513': 328, '514': 329, '515': 330, '516': 331, '517': 332, '518': 333, '519': 334, '52': 335, '520': 336, '521': 337, '522': 338, '523': 339, '524': 340, '527': 341, '528': 342, '529': 343, '53': 344, '530': 345, '531': 346, '532': 347, '533': 348, '534': 349, '535': 350, '536': 351, '537': 352, '54': 353, '540': 354, '542': 355, '543': 356, '550': 357, '552': 358, '553': 359, '555': 360, '556': 361, '557': 362, '558': 363, '560': 364, '562': 365, '564': 366, '565': 367, '566': 368, '567': 369, '568': 370, '569': 371, '570': 372, '571': 373, '572': 374, '573': 375, '574': 376, '575': 377, '576': 378, '577': 379, '578': 380, '579': 381, '580': 382, '581': 383, '583': 384, '584': 385, '585': 386, '586': 387, '588': 388, '590': 389, '591': 390, '592': 391, '593': 392, '594': 393, '595': 394, '596': 395, '598': 396, '599': 397, '600': 398, '601': 399, '602': 400, '603': 401, '604': 402, '605': 403, '607': 404, '608': 405, '610': 406, '611': 407, '614': 408, '615': 409, '616': 410, '617': 411, '618': 412, '619': 413, '620': 414, '621': 415, '622': 416, '623': 417, '625': 418, '626': 419, '627': 420, '634': 421, '641': 422, '642': 423, '644': 424, '645': 425, '646': 426, '647': 427, '648': 428, '649': 429, '652': 430, '654': 431, '656': 432, '658': 433, '659': 434, '66': 435, '661': 436, '663': 437, '664': 438, '665': 439, '670': 440, '674': 441, '680': 442, '681': 443, '682': 444, '683': 445, '684': 446, '685': 447, '686': 448, '691': 449, '692': 450, '693': 451, '694': 452, '695': 453, '696': 454, '698': 455, '7': 456, '70': 457, '701': 458, '702': 459, '703': 460, '704': 461, '705': 462, '706': 463, '707': 464, '709': 465, '710': 466, '711': 467, '712': 468, '713': 469, '714': 470, '715': 471, '716': 472, '717': 473, '718': 474, '719': 475, '721': 476, '722': 477, '723': 478, '724': 479, '725': 480, '726': 481, '727': 482, '728': 483, '729': 484, '730': 485, '731': 486, '733': 487, '734': 488, '736': 489, '737': 490, '738': 491, '741': 492, '742': 493, '745': 494, '746': 495, '747': 496, '748': 497, '75': 498, '750': 499, '751': 500, '752': 501, '753': 502, '754': 503, '755': 504, '756': 505, '758': 506, '759': 507, '78': 508, '780': 509, '781': 510, '782': 511, '783': 512, '784': 513, '785': 514, '786': 515, '787': 516, '788': 517, '789': 518, '79': 519, '790': 520, '791': 521, '792': 522, '793': 523, '794': 524, '795': 525, '796': 526, '797': 527, '799': 528, '8': 529, '800': 530, '801': 531, '802': 532, '805': 533, '806': 534, '807': 535, '808': 536, '810': 537, '811': 538, '812': 539, '813': 540, '814': 541, '815': 542, '816': 543, '820': 544, '821': 545, '822': 546, '823': 547, '824': 548, '825': 549, '826': 550, '831': 551, '832': 552, '833': 553, '836': 554, '837': 555, '840': 556, '842': 557, '843': 558, '844': 559, '845': 560, '846': 561, '847': 562, '850': 563, '851': 564, '852': 565, '853': 566, '860': 567, '861': 568, '862': 569, '863': 570, '864': 571, '865': 572, '866': 573, '867': 574, '868': 575, '869': 576, '870': 577, '871': 578, '872': 579, '873': 580, '879': 581, '88': 582, '880': 583, '881': 584, '882': 585, '883': 586, '884': 587, '891': 588, '892': 589, '893': 590, '894': 591, '9': 592, '905': 593, '906': 594, '907': 595, '908': 596, '909': 597, '910': 598, '911': 599, '912': 600, '913': 601, '915': 602, '916': 603, '917': 604, '918': 605, '919': 606, '920': 607, '921': 608, '922': 609, '923': 610, '924': 611, '927': 612, '933': 613, '934': 614, '94': 615, '942': 616, '944': 617, '945': 618, '947': 619, '948': 620, '952': 621, '953': 622, '955': 623, '958': 624, '959': 625, '96': 626, '962': 627, '963': 628, '965': 629, '967': 630, '968': 631, '969': 632, '972': 633, '974': 634, '975': 635, '977': 636, '980': 637, '987': 638, '989': 639, '99': 640, '990': 641, '991': 642, '992': 643, '994': 644, '995': 645, '996': 646, '997': 647, '998': 648, '999': 649, 'E812': 650, 'E813': 651, 'E814': 652, 'E816': 653, 'E817': 654, 'E818': 655, 'E819': 656, 'E821': 657, 'E826': 658, 'E829': 659, 'E849': 660, 'E850': 661, 'E853': 662, 'E854': 663, 'E858': 664, 'E868': 665, 'E870': 666, 'E878': 667, 'E879': 668, 'E880': 669, 'E881': 670, 'E882': 671, 'E883': 672, 'E884': 673, 'E885': 674, 'E887': 675, 'E888': 676, 'E890': 677, 'E900': 678, 'E905': 679, 'E906': 680, 'E915': 681, 'E916': 682, 'E917': 683, 'E918': 684, 'E919': 685, 'E924': 686, 'E927': 687, 'E928': 688, 'E929': 689, 'E930': 690, 'E931': 691, 'E932': 692, 'E933': 693, 'E934': 694, 'E935': 695, 'E936': 696, 'E937': 697, 'E938': 698, 'E939': 699, 'E941': 700, 'E942': 701, 'E944': 702, 'E945': 703, 'E947': 704, 'E950': 705, 'E965': 706, 'E968': 707, 'E980': 708, 'V02': 709, 'V03': 710, 'V08': 711, 'V09': 712, 'V10': 713, 'V11': 714, 'V12': 715, 'V13': 716, 'V14': 717, 'V15': 718, 'V16': 719, 'V17': 720, 'V18': 721, 'V23': 722, 'V25': 723, 'V42': 724, 'V43': 725, 'V44': 726, 'V45': 727, 'V46': 728, 'V49': 729, 'V50': 730, 'V53': 731, 'V54': 732, 'V55': 733, 'V57': 734, 'V58': 735, 'V60': 736, 'V61': 737, 'V62': 738, 'V63': 739, 'V64': 740, 'V65': 741, 'V66': 742, 'V69': 743, 'V70': 744, 'V72': 745, 'V85': 746, 'V86': 747}
d2_diag_2           =   rev_dict(dict_diag_2)
print(d2_diag_2)

In [ ]:
dict_diag_3             =    {'11': 0, '110': 1, '111': 2, '112': 3, '115': 4, '117': 5, '122': 6, '123': 7, '131': 8, '132': 9, '135': 10, '136': 11, '138': 12, '139': 13, '14': 14, '141': 15, '146': 16, '148': 17, '150': 18, '151': 19, '152': 20, '153': 21, '154': 22, '155': 23, '156': 24, '157': 25, '158': 26, '161': 27, '162': 28, '163': 29, '164': 30, '17': 31, '170': 32, '171': 33, '172': 34, '173': 35, '174': 36, '175': 37, '179': 38, '180': 39, '182': 40, '183': 41, '185': 42, '186': 43, '188': 44, '189': 45, '191': 46, '192': 47, '193': 48, '195': 49, '196': 50, '197': 51, '198': 52, '199': 53, '200': 54, '201': 55, '202': 56, '203': 57, '204': 58, '205': 59, '208': 60, '211': 61, '214': 62, '215': 63, '216': 64, '217': 65, '218': 66, '220': 67, '223': 68, '225': 69, '226': 70, '227': 71, '228': 72, '230': 73, '233': 74, '235': 75, '236': 76, '238': 77, '239': 78, '240': 79, '241': 80, '242': 81, '243': 82, '244': 83, '245': 84, '246': 85, '250': 86, '250.01': 87, '250.02': 88, '250.03': 89, '250.1': 90, '250.11': 91, '250.12': 92, '250.13': 93, '250.2': 94, '250.21': 95, '250.22': 96, '250.23': 97, '250.3': 98, '250.31': 99, '250.4': 100, '250.41': 101, '250.42': 102, '250.43': 103, '250.5': 104, '250.51': 105, '250.52': 106, '250.53': 107, '250.6': 108, '250.7': 109, '250.8': 110, '250.81': 111, '250.82': 112, '250.83': 113, '250.9': 114, '250.91': 115, '250.92': 116, '250.93': 117, '251': 118, '252': 119, '253': 120, '255': 121, '256': 122, '258': 123, '259': 124, '260': 125, '261': 126, '262': 127, '263': 128, '265': 129, '266': 130, '268': 131, '27': 132, '270': 133, '271': 134, '272': 135, '273': 136, '274': 137, '275': 138, '276': 139, '277': 140, '278': 141, '279': 142, '280': 143, '281': 144, '282': 145, '283': 146, '284': 147, '285': 148, '286': 149, '287': 150, '288': 151, '289': 152, '290': 153, '291': 154, '292': 155, '293': 156, '294': 157, '295': 158, '296': 159, '297': 160, '298': 161, '299': 162, '3': 163, '300': 164, '301': 165, '303': 166, '304': 167, '305': 168, '306': 169, '307': 170, '308': 171, '309': 172, '310': 173, '311': 174, '312': 175, '313': 176, '314': 177, '315': 178, '317': 179, '318': 180, '319': 181, '323': 182, '327': 183, '331': 184, '332': 185, '333': 186, '334': 187, '335': 188, '336': 189, '337': 190, '338': 191, '34': 192, '340': 193, '341': 194, '342': 195, '343': 196, '344': 197, '345': 198, '346': 199, '347': 200, '348': 201, '349': 202, '35': 203, '350': 204, '351': 205, '353': 206, '354': 207, '355': 208, '356': 209, '357': 210, '358': 211, '359': 212, '360': 213, '361': 214, '362': 215, '365': 216, '365.44': 217, '366': 218, '368': 219, '369': 220, '370': 221, '372': 222, '373': 223, '374': 224, '376': 225, '377': 226, '378': 227, '379': 228, '38': 229, '380': 230, '381': 231, '382': 232, '383': 233, '384': 234, '385': 235, '386': 236, '387': 237, '388': 238, '389': 239, '391': 240, '394': 241, '395': 242, '396': 243, '397': 244, '398': 245, '401': 246, '402': 247, '403': 248, '404': 249, '405': 250, '41': 251, '410': 252, '411': 253, '412': 254, '413': 255, '414': 256, '415': 257, '416': 258, '417': 259, '42': 260, '420': 261, '421': 262, '423': 263, '424': 264, '425': 265, '426': 266, '427': 267, '428': 268, '429': 269, '430': 270, '431': 271, '432': 272, '433': 273, '434': 274, '435': 275, '436': 276, '437': 277, '438': 278, '440': 279, '441': 280, '442': 281, '443': 282, '444': 283, '445': 284, '446': 285, '447': 286, '448': 287, '451': 288, '452': 289, '453': 290, '454': 291, '455': 292, '456': 293, '457': 294, '458': 295, '459': 296, '460': 297, '461': 298, '462': 299, '463': 300, '464': 301, '465': 302, '466': 303, '47': 304, '470': 305, '472': 306, '473': 307, '475': 308, '477': 309, '478': 310, '480': 311, '481': 312, '482': 313, '483': 314, '484': 315, '485': 316, '486': 317, '487': 318, '49': 319, '490': 320, '491': 321, '492': 322, '493': 323, '494': 324, '495': 325, '496': 326, '5': 327, '500': 328, '501': 329, '506': 330, '507': 331, '508': 332, '510': 333, '511': 334, '512': 335, '514': 336, '515': 337, '516': 338, '517': 339, '518': 340, '519': 341, '521': 342, '522': 343, '523': 344, '524': 345, '525': 346, '527': 347, '528': 348, '529': 349, '53': 350, '530': 351, '531': 352, '532': 353, '533': 354, '534': 355, '535': 356, '536': 357, '537': 358, '538': 359, '54': 360, '540': 361, '542': 362, '543': 363, '550': 364, '552': 365, '553': 366, '555': 367, '556': 368, '557': 369, '558': 370, '560': 371, '562': 372, '564': 373, '565': 374, '566': 375, '567': 376, '568': 377, '569': 378, '57': 379, '570': 380, '571': 381, '572': 382, '573': 383, '574': 384, '575': 385, '576': 386, '577': 387, '578': 388, '579': 389, '580': 390, '581': 391, '582': 392, '583': 393, '584': 394, '585': 395, '586': 396, '588': 397, '590': 398, '591': 399, '592': 400, '593': 401, '594': 402, '595': 403, '596': 404, '597': 405, '598': 406, '599': 407, '600': 408, '601': 409, '602': 410, '603': 411, '604': 412, '605': 413, '607': 414, '608': 415, '610': 416, '611': 417, '614': 418, '616': 419, '617': 420, '618': 421, '619': 422, '620': 423, '621': 424, '622': 425, '623': 426, '624': 427, '625': 428, '626': 429, '627': 430, '641': 431, '642': 432, '643': 433, '644': 434, '646': 435, '647': 436, '648': 437, '649': 438, '652': 439, '653': 440, '654': 441, '655': 442, '656': 443, '657': 444, '658': 445, '659': 446, '66': 447, '660': 448, '661': 449, '663': 450, '664': 451, '665': 452, '669': 453, '670': 454, '671': 455, '674': 456, '680': 457, '681': 458, '682': 459, '684': 460, '685': 461, '686': 462, '690': 463, '692': 464, '693': 465, '694': 466, '695': 467, '696': 468, '697': 469, '698': 470, '7': 471, '70': 472, '701': 473, '702': 474, '703': 475, '704': 476, '705': 477, '706': 478, '707': 479, '708': 480, '709': 481, '710': 482, '711': 483, '712': 484, '713': 485, '714': 486, '715': 487, '716': 488, '717': 489, '718': 490, '719': 491, '720': 492, '721': 493, '722': 494, '723': 495, '724': 496, '725': 497, '726': 498, '727': 499, '728': 500, '729': 501, '730': 502, '731': 503, '732': 504, '733': 505, '734': 506, '735': 507, '736': 508, '737': 509, '738': 510, '741': 511, '742': 512, '744': 513, '745': 514, '746': 515, '747': 516, '75': 517, '750': 518, '751': 519, '752': 520, '753': 521, '754': 522, '755': 523, '756': 524, '757': 525, '758': 526, '759': 527, '78': 528, '780': 529, '781': 530, '782': 531, '783': 532, '784': 533, '785': 534, '786': 535, '787': 536, '788': 537, '789': 538, '79': 539, '790': 540, '791': 541, '792': 542, '793': 543, '794': 544, '795': 545, '796': 546, '797': 547, '799': 548, '8': 549, '800': 550, '801': 551, '802': 552, '805': 553, '807': 554, '808': 555, '810': 556, '811': 557, '812': 558, '813': 559, '814': 560, '815': 561, '816': 562, '820': 563, '821': 564, '822': 565, '823': 566, '824': 567, '825': 568, '826': 569, '831': 570, '834': 571, '836': 572, '837': 573, '838': 574, '840': 575, '841': 576, '842': 577, '844': 578, '845': 579, '847': 580, '848': 581, '850': 582, '851': 583, '852': 584, '853': 585, '854': 586, '860': 587, '861': 588, '862': 589, '863': 590, '864': 591, '865': 592, '866': 593, '867': 594, '868': 595, '870': 596, '871': 597, '872': 598, '873': 599, '875': 600, '876': 601, '877': 602, '879': 603, '88': 604, '880': 605, '881': 606, '882': 607, '883': 608, '884': 609, '890': 610, '891': 611, '892': 612, '893': 613, '9': 614, '905': 615, '906': 616, '907': 617, '908': 618, '909': 619, '910': 620, '911': 621, '912': 622, '913': 623, '915': 624, '916': 625, '917': 626, '918': 627, '919': 628, '920': 629, '921': 630, '922': 631, '923': 632, '924': 633, '928': 634, '930': 635, '933': 636, '934': 637, '935': 638, '94': 639, '942': 640, '943': 641, '944': 642, '945': 643, '948': 644, '951': 645, '952': 646, '953': 647, '955': 648, '956': 649, '958': 650, '959': 651, '962': 652, '965': 653, '966': 654, '967': 655, '969': 656, '970': 657, '971': 658, '972': 659, '980': 660, '987': 661, '989': 662, '991': 663, '992': 664, '995': 665, '996': 666, '997': 667, '998': 668, '999': 669, 'E812': 670, 'E813': 671, 'E815': 672, 'E816': 673, 'E817': 674, 'E818': 675, 'E819': 676, 'E822': 677, 'E825': 678, 'E826': 679, 'E828': 680, 'E849': 681, 'E850': 682, 'E852': 683, 'E853': 684, 'E854': 685, 'E855': 686, 'E858': 687, 'E861': 688, 'E864': 689, 'E865': 690, 'E870': 691, 'E876': 692, 'E878': 693, 'E879': 694, 'E880': 695, 'E881': 696, 'E882': 697, 'E883': 698, 'E884': 699, 'E885': 700, 'E886': 701, 'E887': 702, 'E888': 703, 'E892': 704, 'E894': 705, 'E900': 706, 'E901': 707, 'E904': 708, 'E905': 709, 'E906': 710, 'E912': 711, 'E915': 712, 'E916': 713, 'E917': 714, 'E919': 715, 'E920': 716, 'E922': 717, 'E924': 718, 'E927': 719, 'E928': 720, 'E929': 721, 'E930': 722, 'E931': 723, 'E932': 724, 'E933': 725, 'E934': 726, 'E935': 727, 'E936': 728, 'E937': 729, 'E938': 730, 'E939': 731, 'E941': 732, 'E942': 733, 'E943': 734, 'E944': 735, 'E945': 736, 'E946': 737, 'E947': 738, 'E949': 739, 'E950': 740, 'E955': 741, 'E956': 742, 'E965': 743, 'E966': 744, 'E980': 745, 'E987': 746, 'V01': 747, 'V02': 748, 'V03': 749, 'V06': 750, 'V07': 751, 'V08': 752, 'V09': 753, 'V10': 754, 'V11': 755, 'V12': 756, 'V13': 757, 'V14': 758, 'V15': 759, 'V16': 760, 'V17': 761, 'V18': 762, 'V22': 763, 'V23': 764, 'V25': 765, 'V27': 766, 'V42': 767, 'V43': 768, 'V44': 769, 'V45': 770, 'V46': 771, 'V49': 772, 'V53': 773, 'V54': 774, 'V55': 775, 'V57': 776, 'V58': 777, 'V60': 778, 'V61': 779, 'V62': 780, 'V63': 781, 'V64': 782, 'V65': 783, 'V66': 784, 'V70': 785, 'V72': 786, 'V85': 787, 'V86': 788}
d2_diag_3           =   rev_dict(dict_diag_3)
print(d2_diag_3)

### Cleaned data for model building

In [ ]:
print(df_for_model.shape)
print(df_for_model.head(2).T)

In [ ]:
Null_cols         =   ['race','payer_code','medical_specialty','diag_1','diag_2','diag_3']
dicts             =   [d2_race, d2_payer_code, d2_medical_specialty,d2_diag_1, d2_diag_2,d2_diag_3]
for  i  in  range(5):
    col       =   Null_cols[i]
    dic       =   dicts[i]
    print("\n {}".format(col))
    print("\n {}".format(dic))
    imputed_df.replace({col: dic}, inplace = True)


In [ ]:
Nonnull_Cols_df['race']               =   imputed_df['race']
Nonnull_Cols_df['payer_code']         =   imputed_df['payer_code']
Nonnull_Cols_df['medical_specialty']  =   imputed_df['medical_specialty']
Nonnull_Cols_df['diag_1']             =   imputed_df['diag_1']
Nonnull_Cols_df['diag_2']             =   imputed_df['diag_2']
Nonnull_Cols_df['diag_3']             =   imputed_df['diag_3']

In [ ]:
Nonnull_Cols_df.head(10).T

### Cleaned data for EDA

In [ ]:
csv_file2 = './Data/DSEFTB-JUN24-G2-EDA_' + str(datetime.now().strftime('%Y_%m_%d_%H_%M_%S')) + '.csv'
Nonnull_Cols_df.to_csv(csv_file2, index = False)

In [ ]:
print(Nonnull_Cols_df.shape)

In [ ]:
Nonnull_Cols_df.head(10).T

## END